packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -94,92 +89,87 @@ export default class RequestClient {
+    send and what headers to not send.
+
+    The preflight only happens once throughout the life-cycle of a certain
+    Companion-client <-> Companion-server pair (allowedHeadersCache).
+    Subsequent requests use the cached result of the preflight.
+    However if there is an error retrieving the allowed headers, we will try again next time
+  */
+  async preflight (path) {
+    const allowedHeadersCached = allowedHeadersCache.get(this.hostname)
+    if (allowedHeadersCached != null) return allowedHeadersCached
+
+    const fallbackAllowedHeaders = ['accept', 'content-type', 'uppy-auth-token']
+
+    const promise = (async () => {
+      try {
+        const response = await fetch(this.#getUrl(path), { method: 'OPTIONS' })
+
+        const header = response.headers.get('access-control-allow-headers')
+        if (header == null) {
+          allowedHeadersCache.set(this.hostname, fallbackAllowedHeaders)
+          return undefined
+        }
 
-  preflight (path) {
-    if (this.preflightDone) {
-      return Promise.resolve(this.allowedHeaders.slice())
-    }
+        this.uppy.log(`[CompanionClient] adding allowed preflight headers to companion cache: ${this.hostname} ${header}`)
 
-    return fetch(this.#getUrl(path), {
-      method: 'OPTIONS',
-    })
-      .then((response) => {
-        if (response.headers.has('access-control-allow-headers')) {
-          this.allowedHeaders = response.headers.get('access-control-allow-headers')
-            .split(',').map((headerName) => headerName.trim().toLowerCase())
-        }
-        this.preflightDone = true
-        return this.allowedHeaders.slice()
-      })
-      .catch((err) => {
+        const allowedHeaders = header.split(',').map((headerName) => headerName.trim().toLowerCase())
+        allowedHeadersCache.set(this.hostname, allowedHeaders)
+        return allowedHeaders
+      } catch (err) {
         this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
-        this.preflightDone = true
-        return this.allowedHeaders.slice()
-      })
-  }
+        // If the user gets a network error or similar, we should try preflight
+        // again next time, or else we might get incorrect behaviour.
+        allowedHeadersCache.delete(this.hostname)
+        return undefined
+      }
+    })()
 
-  preflightAndHeaders (path) {
-    return Promise.all([this.preflight(path), this.headers()])
-      .then(([allowedHeaders, headers]) => {
-        // filter to keep only allowed Headers
-        Object.keys(headers).forEach((header) => {
-          if (!allowedHeaders.includes(header.toLowerCase())) {
-            this.uppy.log(`[CompanionClient] excluding disallowed header ${header}`)
-            delete headers[header] // eslint-disable-line no-param-reassign
-          }
-        })
-
-        return headers
-      })
+    allowedHeadersCache.set(this.hostname, promise)
+    const allowedHeadersNew = await promise
```

Comment:
was there a reason that we set the unresolved promise to the cache instead of the resolved one?

Reply:
yes, it prevents the race condition we're trying to fix. we also set the resolved one later. if we didn't set the cache to the unresolved promise, the next immeidate call would not see it

Reply:
Caching resolved promise is very bad for garbage collecting anyway, it's (almost?) always better to cache the value it resolves to than the promise.